libnetdata/popen/popen.c

```diff
@@ -91,23 +95,36 @@ static inline FILE *custom_popene(const char *command, volatile pid_t *pidptr, c
     posix_spawnattr_t attr;
     posix_spawn_file_actions_t fa;
 
-    if (pipe(pipefd) == -1)
-        return NULL;
-    if ((fp = fdopen(pipefd[PIPE_READ], "r")) == NULL) {
-        goto error_after_pipe;
+    if (flags & FLAG_CREATE_PIPE) {
+        if (pipe(pipefd) == -1)
+            return NULL;
+        if ((fp = fdopen(pipefd[PIPE_READ], "r")) == NULL) {
+            goto error_after_pipe;
+        }
     }
 
-    // Mark all files to be closed by the exec() stage of posix_spawn()
-    int i;
-    for (i = (int) (sysconf(_SC_OPEN_MAX) - 1); i >= 0; i--)
-        if(i != STDIN_FILENO && i != STDERR_FILENO)
-            (void)fcntl(i, F_SETFD, FD_CLOEXEC);
+    if (flags & FLAG_CLOSE_FD) {
+        // Mark all files to be closed by the exec() stage of posix_spawn()
+        int i;
+        for (i = (int) (sysconf(_SC_OPEN_MAX) - 1); i >= 0; i--) {
+            if (i != STDIN_FILENO && i != STDERR_FILENO)
+                (void) fcntl(i, F_SETFD, FD_CLOEXEC);
+        }
+    }
 
     if (!posix_spawn_file_actions_init(&fa)) {
-        // move the pipe to stdout in the child
-        if (posix_spawn_file_actions_adddup2(&fa, pipefd[PIPE_WRITE], STDOUT_FILENO)) {
-            error("posix_spawn_file_actions_adddup2() failed");
-            goto error_after_posix_spawn_file_actions_init;
+        if (flags & FLAG_CREATE_PIPE) {
+            // move the pipe to stdout in the child
+            if (posix_spawn_file_actions_adddup2(&fa, pipefd[PIPE_WRITE], STDOUT_FILENO)) {
+                error("posix_spawn_file_actions_adddup2() failed");
+                goto error_after_posix_spawn_file_actions_init;
+            }
+        } else {
+            // set stdout to /dev/null
+            if (posix_spawn_file_actions_addopen(&fa, STDOUT_FILENO, "/dev/null", O_WRONLY, 0)) {
+                error("posix_spawn_file_actions_addopen() failed");
+                // this is not a fatal error
+            }
         }
     } else {
```

Comment:
We don't need an if-then-else here when the check bails to an exit, would be cleaner to do:
```
if(posix_spawn_file_actions_init ... ) {
  error(...)
  goto ....
}
// Continue other cases
```
I think it makes it easier to read, not an essential change.

Reply:
Agreed. I kept the existing structure here, didn't want to refactor the existing code too much.